### ISSUE TITLE
Cleanup Requested: Examples with Console Warnings

### ISSUE THREAD

**Issue by WestLangley** [mrdoob/three.js#15468]
The following webGL examples throw console warnings:

- `webgl_materials_variations_toon.html` - Output of vertex shader 'vNormal' not read by fragment shader (caused by `effects/OutlineEffect.js)`. Also `webgl_loader_mmd.html`, `webgl_loader_mmd_pose.html`, `webgl_loader_mmd_audio.html` /ping @takahirox

fixed:

- ~~`webgl_loader_kmz.html` - Missing doc.kml file /ping @mrdoob~~

- ~~`webgl_loader_obj2_options.html` - object_group "_default" was defined with unresolvable material /ping @kaisalmen~~

- ~~`webgl_materials_nodes.html` - Output of vertex shader 'vObjectNormal' not read by fragment shader... additional errors with other dropdown selections /ping @sunag~~

- ~~`webgl_mirror_nodes.html` - Output of vertex shader 'vObjectNormal' not read by fragment shader /ping @sunag~~

- ~~`webgl_sprites_nodes.html` - sprites nodes - Output of vertex shader 'vPosition' not read by fragment shader /ping @sunag~~

- ~~`webgl_loader_sea3d_morph.html` - image is not power of two (200x200) /ping @sunag~~

- ~~`webgl_loader_ctm_materials.html` - .metal has been removed~~

- ~~`webgl_materials_bumpmap_skin.html` - Output of vertex shader 'vDirectionalShadowCoord' not read by fragment shader~~

- ~~`webgl_panorama_dualfisheye.html` - image is not power of two (1920x1090)~~

- ~~`webgl_postprocessing_masking.html` - image is not power of two (758x600)~~

- ~~`webgl_shaders_ocean2.html` - Output of vertex shader 'vUV' not read by fragment shader~~

- ~~`webgl_loader_gcode.html` - Command not supported~~

three.js `105dev`

**Comment by sunag** [comment]
Hi @WestLangley 
For exception of `webgl_loader_sea3d_morph.html` I not found no warnings in node examples.
this has already been fixed?

**Comment by sunag** [comment]
Hmm. I download the zip for warranty but still no warnings alert. Maybe my browser version or it is a gpu card alert? I test in chrome v7.1, v7.3 and firefox
![image](https://user-images.githubusercontent.com/502810/50453173-ead86180-0925-11e9-8776-fcea392638b7.png)

**Comment by WestLangley** [comment]
Hmm... Let's see what others observe...

**Comment by Mugen87** [comment]
> Hmm... Let's see what others observe...

I've tested all listed examples and I see the same warnings @WestLangley mentioned in his first post.

**Comment by WestLangley** [comment]
@sunag Are you still unable to see the warnings in your examples?

**Comment by titansoftime** [comment]
Not sure if this is related, but I get this warning:

`THREE.WebGLRenderer: WEBGL_compressed_texture_pvrtc extension not supported.`

I am loading .dds (bc1 and bc3) textures using the DDSLoader and have no issues whatsoever.

**Comment by WestLangley** [comment]
@titansoftime This issue is in regards to the three.js examples that throw console warnings. Which three.js example are you referring to?

**Comment by titansoftime** [comment]
Ah, I see. I do not believe there are any examples demonstrating the usage of compressed textures, so nevermind =]

**Comment by WestLangley** [comment]
@titansoftime 

https://threejs.org/examples/webgl_loader_texture_ktx.html
https://threejs.org/examples/webgl_loader_texture_pvrtc.html

@Mugen87 Do you know what is going on with these? Maybe they need to fail more gracefully?

**Comment by Mugen87** [comment]
Not all texture compression formats are supported on all hardware. For instance desktops do not support PVRTC since it's intended for iOS hardware. When `three.js` checks for available texture compression formats, it does it in a certain order, see

https://github.com/mrdoob/three.js/blob/d6158ec0cbfe0b882ff744f8586774907f0c434a/src/renderers/webgl/WebGLState.js#L489-L492

That's the reason why you see the warning in [this example](https://threejs.org/examples/webgl_loader_texture_dds) where no `PVRTC` is used (but `S3TC`). An engine can only reliably test for hardware support if it tries to get the respective extension over the WebGL context. No matter what order you choose, you will always see these warnings in the console.

**Comment by WestLangley** [comment]
Thanks, @Mugen87. Can we test for the user-agent and suppress unnecessary warnings that way?

**Comment by Mugen87** [comment]
Even if you detect the browser and OS, you still don't know if the actual device supports a given texture compression format. AFAIK, the only way of doing this is the request for the respective WebGL extension.

**Comment by kaisalmen** [comment]
I have finally addressed the issue of `webgl_loader_obj2_options.html` in the referenced PR above.

**Comment by WestLangley** [comment]
@mrdoob It has been a year. Should I close this?

**Comment by WestLangley** [comment]
@takahirox Will you please address the examples mentioned in the first post?

**Comment by looeee** [comment]
> webgl_materials_variations_toon.html - Output of vertex shader 'vNormal' not read by fragment shader (caused by effects/OutlineEffect.js). Also webgl_loader_mmd.html, webgl_loader_mmd_pose.html, webgl_loader_mmd_audio.html

The only one of these I get warning for is _webgl_loader_mmd_audio.html_:

``` 
The AudioContext was not allowed to start. It must be resumed (or created) after a user gesture on the page
```

No warnings for the others, tested on latest Chrome in Windows.

**Comment by takahirox** [comment]
Sorry I didn't notice this thread. I don't see the warning other than webgl_loader_mmd_audio example neither on my Windows Chrome.

Regarding webgl_loader_mmd_audio, that warning makes sense due to the security policy update. I'll fix it or at least open an issue, so you can close this issue if you don't see the warning in other examples.

**Comment by WestLangley** [comment]
With macOS, I get the errors with Chrome, Safari, and Firefox.

    WARNING: Output of vertex shader 'vNormal' not read by fragment shader

    WARNING: Output of vertex shader 'webgl_d63a0821229be6c6' not read by fragment shader

    WARNING: Output of vertex shader 'vViewPosition' not read by fragment shader

**Comment by WestLangley** [comment]
As I said in the top post, check the shader created by `OutlineEffect.js`.

**Comment by looeee** [comment]
Checked in Firefox and Edge on Windows for good measure. No warning other than the audio one there either.

It seems like this is either a MacOS issue or a GPU issue. Can someone else with MacOS check to see if they get the warnings?

**Comment by sciecode** [comment]
I can confirm what @WestLangley mentioned, MacOS with integrated Intel GPU.

Chrome - only `vNormal` and `webgl_d63a0821229be6c6`
Firefox - only `webgl_d63a0821229be6c6`
Safari -  only `vViewPosition`

Very curious.

**Comment by WestLangley** [comment]
>It seems like this is either a MacOS issue or a GPU issue.

No. Look at the vertex and fragment shaders created by `OutlineEffect` and see if the defined varyings are used in the fragment shader.

```glsl
#ifndef FLAT_SHADED
	varying vec3 vNormal;
#endif

...

varying vec3 vViewPosition;
```

**Comment by looeee** [comment]
@WestLangley do you have any idea why these warnings seem to only occur on MacOS?

**Comment by WestLangley** [comment]
>@WestLangley do you have any idea why these warnings seem to only occur on MacOS?

@looeee The warnings also appear for me on iOS Safari.

**Comment by looeee** [comment]
OK, thanks for checking iOS as well. The question still stands though - why are these warnings only occurring on some operating systems (perhaps just Apple's) and not others? Is this caused by a missing extension? 

Also, @takahirox won't be able to fix the warnings if he can't see them. Someone with Apple hardware will have to take this on.

**Comment by WestLangley** [comment]
>Someone with Apple hardware will have to take this on.

@looeee The author of `OutlineEffect.js` needs to avoid generating a shader in which varyings are declared in the vertex shader that are not read by the fragment shader. I don't think Apple hardware is required to make that change.

**Comment by WestLangley** [comment]
TL;DR

If someone is interested in pursuing this, `OutlineEffect.js` declares `vNormal` and `vViewPosition`  in the vertex shader, but does not use them in the fragment shader.

The following examples are affected:

    webgl_materials_variations_toon.html
    webgl_loader_mmd.html
    webgl_loader_mmd_pose.html
    webgl_loader_mmd_audio.html
